feat: Parse json or js config

joshuef opened the pull request:
this allows to use a javasript config file as well maintaining json config file functionality.

This gives a bit more flexibility to app developers (we use for setting vars depending on if testing packed or unpacked electron apps)

joshuef commented:
Let me know if any tweaks or changes are needed to accept this 👍

AndreyBelym commented on src/utils/get-config.js:
```diff
@@ -12,9 +12,18 @@ export default function (id, mainPath) {
-    var configString = readFileSync(mainPath).toString();
 
-    var config = JSON.parse(configString);
+    var config;
```
I think we can do it in a more elegant way:
```js
const configExtension = path.extname(mainPath);

if (Object.keys(require.extensions).some(extention => extension  === configExtension)) 
    config = require(mainPath);
else {
    var configString = readFileSync(mainPath).toString();

    config = JSON.parse(configString);
}
```

This way we won't lose an error if it happens during `require`.

joshuef replied:
Okay rad, I'll update the code and look to making a test for this 👍

joshuef replied:
This doesn't work as `extension  === configExtension` is never met... As there's no extension on our configExtension (just `testcafe-electron-rc`). 

Hmmm, the original code just tries for _anything_ resolveable first, and on failure, fallsback to json... I'm not sure what we want to be doing with a require error there, as we're falling back. Perhaps we just want to print a warning so folk are aware?

joshuef replied:
```js

    try {
        config = require(mainPath);
    }
    catch ( e ) {

        if ( !e.message.includes('Cannot find module') )

            throw e;

        var configString = readFileSync(mainPath).toString();

        config = JSON.parse(configString);
    }
```

AndreyBelym replied:
`require` won't load a JS config if its name doesn't end with `.js`. So, if you want to use JS, you *will have* to explicitly specify the `.js` extension.

Honestly, I consider this naming decision (`.testcafe-electron-rc` without the explicit `.json` extension) as my mistake. If you explicitly specify an extension, you will get at least basic autocompletion in IDEs like VS Code and WebStorm. And maybe someday we will add a JSON schema and submit it to https://schemastore.org to get the full autocompletion.

So, I still prefer the way I described. In my opinion, using explicit extensions is the right way, and if there is no extension or if it is an unknown one, we should try the current legacy way with file reading and JSON parsing.

joshuef replied:
>if there is no extension or if it is an unknown one, we should try the current legacy way with file reading and JSON parsing.
But we can try both `js` and default to `json` when using a directory too. I don't see why _not_ to try requiring `.js` files (or other supported formats in future)...

I had toyed with the idea of looping over `require.extensions` to do just that, but actually [it seems like we should _not_ be using this, as it's been marked `deprecated` in latest node :( ](https://nodejs.org/api/modules.html#modules_require_extensions)

AndreyBelym replied:
I understand that it seems dangerous using a `deprecated` functionality. However, IMHO it is deprecated since it is a very low-level feature, rather than an obsolete one. There is no other way to control `require`'s behaviour for different file types, so even big projects like `ts-node` use it. Moreover, the TestCafe's compiler subsystem heavily relies on this feature. However, I won't insist if you don't want to play with it 😄

joshuef replied:
I can re add this if you want to be using it. I just think it may be a headache for removing it later. 

I'm also of the opinion that _at least_ `.js` files should be detected and as seamless as the base `testcafe-electron-rc` (ie, we shouldn't _need_ to specify it when running tests). So if you're happy doing that, it probably makes sense to do that for _all_ extensions present in `require.extensions`, no? 

Let me know what you think and I can get the PR updated :+1:

joshuef commented:
I've updated with a test (though couldn't get them to run locally, so let's see what CI says), and updated to throw any non 'not found' errors on the require attempt. 

Let me know if more changes are needed 🙇

AndreyBelym commented on Gulpfile.js:
```diff
@@ -16,6 +16,7 @@ var APP_DIR             = path.join(__dirname, 'test/data/test-app-regular');
 var ASAR_ARCHIVE_PATH   = path.join(__dirname, 'test/data/test-app.asar');
 var CONFIG_PATH_REGULAR = path.join(__dirname, 'test/data/app-config-regular');
 var CONFIG_PATH_ASAR    = path.join(__dirname, 'test/data/app-config-asar');
+var CONFIG_PATH_JS      = path.join(__dirname, 'test/data/app-config-js');
```
`require` won't be able to load a module without the correct extension (`.js`). I've added a more detailed comment to our existing conversation: https://github.com/DevExpress/testcafe-browser-provider-electron/pull/54#discussion_r361151757

joshuef commented:
I've updated the code for a json/js check, and commented above re: the suggested solution.

Tests are going now (locally too, things work better when you don't have a broken `yarn` install it seems :| 🤦‍♂  ).

joshuef commented:
I've updated the code for a json/js check, and commented above re: the suggested solution.

Tests are going now (locally too, things work better when you don't have a broken `yarn` install it seems :| 🤦‍♂  ).

AndreyBelym commented on src/utils/get-config.js:
```diff
@@ -8,13 +8,31 @@ import CONSTANTS from '../constants';
-    var config = JSON.parse(configString);
+        if ( !e.message.includes('Cannot find module') )
+
+            throw e;
```
I'm still not fond of masking the `Cannot find module` issue. The first problem is that the error message can be changed without a notice in the next Node.js version. The second is that you can specify a full path to a JS config and make a mistake in it. If we mask the error, it can be very hard to detect a misprint without a proper error message.

If you don't want to touch `require.extensions`, we can define a constant list that contains `.js` and `.json` extensions (`require` can load JSON files too), and use `require` if we can detect a supported extension in a config path, and fallback to the legacy loading mechanism otherwise.

joshuef replied:
> The first problem is that the error message can be changed without a notice in the next Node.js version
This seems at odds with the dependency on code marked as deprecated to me.
>The second is that you can specify a full path to a JS config and make a mistake in it
Good point.
>and use require if we can detect a supported extension in a config path, and fallback to the legacy loading mechanism otherwise.
This sounds like best of both worlds 👍

joshuef commented:
Okay, followed your suggestion [here](https://github.com/DevExpress/testcafe-browser-provider-electron/pull/54#discussion_r364731240) @AndreyBelym , we can auto-detect if a `.js` or `.json` file exists (only when a dir is provided).

And if we have that, or some _other_ `extname`, we'll try and `require` the file (w/o skipping any errors). If no `extname` is present on our config (either from passing it in, or detecting our acceptable defaults), we fallback to the `.testcafe-electron-rc` file.

I've also added a test in for passing the config file _specifically_ too.

So I _think_ this is looking like a much nicer solution. Let me know!

joshuef commented:
Hey @AndreyBelym, just wanted to bump this and see if you've any more feedback?

AndreyBelym commented:
I apologize for the delays. I'm happy to merge your PR, thank you for your contribution. However, we will need to change our docs slightly before releasing and publishing it on NPM.

joshuef commented:
No worries about delays @AndreyBelym ! I know how it goes :) 

Thanks for the feedback and help on this. Keep up the good work 👍
